singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,250 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pinterest.singer.loggingaudit.client;
+
+import com.pinterest.singer.loggingaudit.client.common.LoggingAuditClientMetrics;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditHeaders;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditEvent;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditStage;
+import com.pinterest.singer.loggingaudit.thrift.configuration.KafkaSenderConfig;
+import com.pinterest.singer.metrics.OpenTsdbMetricConverter;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.thrift.TException;
+import org.apache.thrift.TSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class AuditEventKafkaSender implements LoggingAuditEventSender {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuditEventKafkaSender.class);
+
+  private final LoggingAuditStage stage;
+  private final String host;
+  private final LinkedBlockingDeque<LoggingAuditEvent> queue;
+  private KafkaProducer<byte[], byte[]> kafkaProducer;
+  private TSerializer serializer = new TSerializer();
+  private AtomicBoolean cancelled = new AtomicBoolean(false);
+  private String topic;
+  private String name;
+  private Thread thread;
+  private int dequeueWaitInSeconds = 30;
+  private int stopGracePeriodInSeconds = 300;
+  private int callingThreadSleepInSeconds = 30;
+  private int numOfPartitionsToTrySending = 3;
+  private Map<Integer, AtomicInteger> partitionErrorCounts = new ConcurrentHashMap<>();
+  private Map<LoggingAuditHeaders, Set<Integer>> eventBadPartitionsMap = new ConcurrentHashMap<>();
+
+  public AuditEventKafkaSender(KafkaSenderConfig config,
+                               LinkedBlockingDeque<LoggingAuditEvent> queue,
+                               LoggingAuditStage stage , String host, String name){
+     this.topic = config.getTopic();
+     this.queue = queue;
+     this.stage = stage;
+     this.host = host;
+     this.name = name;
+  }
+
+  public KafkaProducer<byte[], byte[]> getKafkaProducer() {
+    return kafkaProducer;
+  }
+
+  public void setKafkaProducer(KafkaProducer<byte[], byte[]> kafkaProducer) {
+    this.kafkaProducer = kafkaProducer;
+  }
+
+
+  public int getGoodPartition(Set<Integer> badPartitions, int numOfPartitions){
```

Comment:
`alternatePartition` or `randomPartition` ? This method doesn't check if a partition is good or bad.

Reply:
Good means potentially good. we just want a different partition that is not tried before